### PR TITLE
[DDW-1100] Implement Yoroi wallets restoration

### PR DESCRIPTION
**Pull request by DeeJayElly**
This PR updates Restore Wallet dialog with Yoroi wallets restoration.

### English
Screenshot 2019-12-11 17 31 26

### Japanese
Screenshot 2019-12-11 17 32 01

---

## Testing Checklist

- [Slack QA thread](https://input-output-rnd.slack.com/archives/GGKFXSKC6/p1575585892268700)

---

### Basics

- [ ] PR has been assigned and has appropriate labels (`feature`/`bug`/`chore`, `release-x.x.x`)
- [ ] PR is updated to the most recent version of the target branch (and there are no conflicts)
- [ ] PR has a good description that summarizes all changes and shows some screenshots or animated GIFs of important UI changes
- [ ] CHANGELOG entry has been added to the top of the appropriate section (*Features*, *Fixes*, *Chores*) and is linked to the correct PR on GitHub
- [ ] Automated tests: All acceptance and unit tests are passing (`yarn test`)
- [ ] Manual tests (minimum tests should cover newly added feature/fix): App works correctly in *development* build (`yarn dev`)
- [ ] Manual tests (minimum tests should cover newly added feature/fix): App works correctly in *production* build (`yarn package` / CI builds)
- [ ] There are no *flow* errors or warnings (`yarn flow:test`)
- [ ] There are no *lint* errors or warnings (`yarn lint`)
- [ ] There are no *prettier* errors or warnings (`yarn prettier:check`)
- [ ] There are no missing translations (running `yarn manage:translations` produces no changes)
- [ ] Text changes are proofread and approved (Jane Wild / Amy Reeve)
- [ ] Japanese text changes are proofread and approved (Junko Oda)
- [ ] UI changes look good in all themes (Alexander Rukin)
- [ ] Storybook works and no stories are broken (`yarn storybook`)
- [ ] In case of dependency changes `yarn.lock` file is updated

### Code Quality
- [ ] Important parts of the code are properly commented and documented
- [ ] Code is properly typed with flow
- [ ] React components are split-up enough to avoid unnecessary re-renderings
- [ ] Any code that only works in main process is neatly separated from components

### Testing
- [ ] New feature/change is covered by acceptance tests
- [ ] New feature/change is manually tested and approved by QA team
- [ ] All existing acceptance tests are still up-to-date
- [ ] New feature/change is covered by Daedalus Testing scenario
- [ ] All existing Daedalus Testing scenarios are still up-to-date

### After Review
- [ ] Merge the PR
- [ ] Delete the source branch
- [ ] Move the ticket to `done` column on the YouTrack board
- [ ] Update Slack QA thread by marking it with a green checkmark

### CODE REVIEW

**Comment by IuliaDolishniak** [comment]
@DeeJayElly hi, these could be an expected issues but just wanted to make a note of it:
1. new label on your screen have a different color
2. this is a known issue but still there - in Japanese version we have textboxes on a different height

**Comment by nikolaglumac** [comment]
Issue 1. should be fixed.
Issue 2. is a known issue which we will not be fixing now.
cc @DeeJayElly @IuliaDolishniak

File: source/renderer/app/i18n/locales/en-US.json [hunk 1]
```diff
@@ -132,7 +132,7 @@
   "global.language.german": "German",
   "global.language.japanese": "Japanese / 日本語",
   "global.language.korean": "Korean",
-  "global.passwordInstructions": "The password needs to be at least <strong>10 characters</strong> long, and have at least <strong>1 uppercase letter</strong>, <strong>1 lowercase letter</strong> and <strong>1 number</strong>.",
+  "global.passwordInstructions": "The password needs to be at least <strong>7 characters</strong> long, and have at least <strong>1 uppercase letter</strong>, <strong>1 lowercase letter</strong> and <strong>1 number</strong>.",
```
Comment (nikolaglumac) [inline source/renderer/app/i18n/locales/en-US.json]: @DeeJayElly please change "7" to "10" characters long.
@a-rukin this was wrongly presented in the designs - please update the designs!

File: source/renderer/app/config/walletsConfig.js [hunk 2]
```diff
@@ -21,10 +22,14 @@ export const WALLET_RESTORE_TYPES = {
   REGULAR: 'regular', // Shelley wallet
   CERTIFICATE: 'certificate', // Paper wallet
   LEGACY: 'legacy', // Byron wallet
+  YOROI_REWARDS: 'yoroi-rewards', // Yoroi rewards wallet
```
Comment (nikolaglumac) [inline source/renderer/app/config/walletsConfig.js]: @DeeJayElly please use `YOROI_REGULAR: 'yoroi-regular'`

File: source/renderer/app/config/walletsConfig.js [hunk 3]
```diff
@@ -21,10 +22,14 @@ export const WALLET_RESTORE_TYPES = {
...
+  YOROI_BALANCE: 'yoroi-balance', // Yoroi balance wallet
```
Comment (nikolaglumac) [inline source/renderer/app/config/walletsConfig.js]: @DeeJayElly please use `YOROI_LEGACY: 'yoroi-legacy'`

File: source/renderer/app/components/wallet/WalletRestoreDialog.scss [hunk 4]
```diff
@@ -15,6 +15,20 @@
     margin-bottom: 20px;
   }
 
+  .newLabel {
+    background-color: var(--theme-main-body-background-color);
```
Comment (nikolaglumac) [inline source/renderer/app/components/wallet/WalletRestoreDialog.scss]: @DeeJayElly please add separate CSS variable for this background color

File: source/renderer/app/components/wallet/WalletRestoreDialog.scss [hunk 5]
```diff
@@ -15,6 +15,20 @@
...
+    border-radius: 3px;
+    color: var(--theme-main-body-messages-color);
```
Comment (nikolaglumac) [inline source/renderer/app/components/wallet/WalletRestoreDialog.scss]: @DeeJayElly please add separate CSS variable for this color

File: source/renderer/app/components/wallet/WalletRestoreDialog.js [hunk 6]
```diff
@@ -551,6 +628,21 @@ export default class WalletRestoreDialog extends Component<Props, State> {
...
+  }
+
+  isYoroi() {
+    return (
+      this.state.walletType === WALLET_RESTORE_TYPES.YOROI_REWARDS ||
+      this.state.walletType === WALLET_RESTORE_TYPES.YOROI_BALANCE
+    );
+  }
+
+  isYoroiRewards() {
```
Comment (nikolaglumac) [inline source/renderer/app/components/wallet/WalletRestoreDialog.js]: @DeeJayElly please rename to `isYoroiRegular()`

File: source/renderer/app/components/wallet/WalletRestoreDialog.js [hunk 7]
```diff
@@ -551,6 +628,21 @@ export default class WalletRestoreDialog extends Component<Props, State> {
     return this.state.walletType === WALLET_RESTORE_TYPES.LEGACY;
   }
 
+  isYoroiBalance() {
```
Comment (nikolaglumac) [inline source/renderer/app/components/wallet/WalletRestoreDialog.js]: @DeeJayElly please rename to `isYoroiLegacy()`

Comment (DeeJayElly) [reply source/renderer/app/components/wallet/WalletRestoreDialog.js]: Done

File: source/renderer/app/components/wallet/WalletRestoreDialog.js (hunk 6 above)
Comment (DeeJayElly) [reply source/renderer/app/components/wallet/WalletRestoreDialog.js]: Done

File: source/renderer/app/config/walletsConfig.js (hunk 3 above)
Comment (DeeJayElly) [reply source/renderer/app/config/walletsConfig.js]: Done

File: source/renderer/app/config/walletsConfig.js (hunk 2 above)
Comment (DeeJayElly) [reply source/renderer/app/config/walletsConfig.js]: Done

File: source/renderer/app/i18n/locales/en-US.json (hunk 1 above)
Comment (DeeJayElly) [reply source/renderer/app/i18n/locales/en-US.json]: Fixed

File: source/renderer/app/components/wallet/WalletRestoreDialog.js [hunk 8]
```diff
@@ -539,10 +620,6 @@ export default class WalletRestoreDialog extends Component<Props, State> {
     );
   }
 
-  isRegular() {
```
Comment (nikolaglumac) [inline source/renderer/app/components/wallet/WalletRestoreDialog.js]: @DeeJayElly why was this removed?

File: source/renderer/app/components/wallet/WalletRestoreDialog.js [hunk 9]
```diff
@@ -474,6 +505,59 @@ export default class WalletRestoreDialog extends Component<Props, State> {
...
 
+        {!this.isCertificate() && this.isYoroi() && (
+          <RadioSet
+            label={intl.formatMessage(messages.recoveryPhraseTypeLabel)}
+            items={[
+              {
+                key: WALLET_RESTORE_TYPES.YOROI_LEGACY,
+                label: (
+                  <Fragment>
+                    15
```
Comment (nikolaglumac) [inline source/renderer/app/components/wallet/WalletRestoreDialog.js]: @DeeJayElly please replace all these hardcoded "number of words" with constants from here: https://github.com/input-output-hk/daedalus/blob/ff8b4559c2bc5473bbb5a322a1f626c7d92307a9/source/renderer/app/config/cryptoConfig.js
(please do that for ALL of them - 12, 15 & 27 words accross this file - and anywhere else where they might still be hardcoded. Thanks!

File: source/renderer/app/components/wallet/WalletRestoreDialog.js [hunk 10]
```diff
@@ -313,6 +322,12 @@ export default class WalletRestoreDialog extends Component<Props, State> {
 
         walletData.type = this.state.walletType;
 
+        if (walletData.type === WALLET_RESTORE_TYPES.YOROI_REGULAR) {
+          walletData.type = WALLET_RESTORE_TYPES.REGULAR;
+        } else if (walletData.type === WALLET_RESTORE_TYPES.YOROI_LEGACY) {
```
Comment (nikolaglumac) [inline source/renderer/app/components/wallet/WalletRestoreDialog.js]: @DeeJayElly this is no good. We should not mask the `YOROI_LEGACY` as a normal `LEGACY` wallet. This info is crucial to know what type of a wallet we are restoring in the Store.

File: source/renderer/app/components/wallet/WalletRestoreDialog.js [hunk 11]
```diff
@@ -380,14 +395,19 @@ export default class WalletRestoreDialog extends Component<Props, State> {
 
     const regularTabClasses = classnames([
       'regularTab',
-      !this.isCertificate() ? styles.activeButton : '',
+      !this.isYoroi() && !this.isCertificate() ? styles.activeButton : '',
```
Comment (nikolaglumac) [inline source/renderer/app/components/wallet/WalletRestoreDialog.js]: @DeeJayElly please replace `!this.isYoroi() && !this.isCertificate()` with `this.isRegular() || this.isLegacy()`

File: source/renderer/app/components/wallet/WalletRestoreDialog.js [hunk 12]
```diff
@@ -380,14 +395,19 @@ export default class WalletRestoreDialog extends Component<Props, State> {
...
     ]);
 
     const certificateTabClasses = classnames([
       'certificateTab',
       this.isCertificate() ? styles.activeButton : '',
     ]);
 
+    const yoroiTabClasses = classnames([
+      'yoroiTab',
+      this.isYoroi() && !this.isCertificate() ? styles.activeButton : '',
```
Comment (nikolaglumac) [inline source/renderer/app/components/wallet/WalletRestoreDialog.js]: @DeeJayElly here it should be enough to use `this.isYoroi()` - I don't think we need the `&& !this.isCertificate()` part...

File: source/renderer/app/components/wallet/WalletRestoreDialog.js [hunk 13]
```diff
@@ -474,6 +505,59 @@ export default class WalletRestoreDialog extends Component<Props, State> {
           />
         )}
 
+        {!this.isCertificate() && this.isYoroi() && (
```
Comment (nikolaglumac) [inline source/renderer/app/components/wallet/WalletRestoreDialog.js]: @DeeJayElly here we should just use `this.isYoroi()`.

File: source/renderer/app/components/wallet/WalletRestoreDialog.js (hunk 11 above)
Comment (DeeJayElly) [reply source/renderer/app/components/wallet/WalletRestoreDialog.js]: Fixed

File: source/renderer/app/components/wallet/WalletRestoreDialog.js (hunk 10 above)
Comment (DeeJayElly) [reply source/renderer/app/components/wallet/WalletRestoreDialog.js]: Fixed

File: source/renderer/app/components/wallet/WalletRestoreDialog.js (hunk 12 above)
Comment (DeeJayElly) [reply source/renderer/app/components/wallet/WalletRestoreDialog.js]: Fixed

File: source/renderer/app/components/wallet/WalletRestoreDialog.js (hunk 13 above)
Comment (DeeJayElly) [reply source/renderer/app/components/wallet/WalletRestoreDialog.js]: Fixed

**Comment by a-rukin** [comment]
@DeeJayElly can you update styles for "new" badge to match this - https://zpl.io/29Ej3Ge

**Comment by a-rukin** [comment]
image
pls also check vertical alignment

**Comment by DeeJayElly** [comment]
> image
> 
> pls also check vertical alignment

Fixed

File: source/renderer/app/components/wallet/WalletRestoreDialog.js (hunk 8 above)
Comment (DeeJayElly) [reply source/renderer/app/components/wallet/WalletRestoreDialog.js]: Reverted back

File: source/renderer/app/components/wallet/WalletRestoreDialog.js (hunk 9 above)
Comment (DeeJayElly) [reply source/renderer/app/components/wallet/WalletRestoreDialog.js]: Fixed

File: source/renderer/app/components/wallet/WalletRestoreDialog.scss (hunk 5 above)
Comment (DeeJayElly) [reply source/renderer/app/components/wallet/WalletRestoreDialog.scss]: Added

File: source/renderer/app/components/wallet/WalletRestoreDialog.scss (hunk 4 above)
Comment (DeeJayElly) [reply source/renderer/app/components/wallet/WalletRestoreDialog.scss]: Added

**Comment by DeeJayElly** [comment]
All changes implemented

**Comment by IuliaDolishniak** [comment]
@DeeJayElly the following error is shown when I enter Daedalus 15 mnemonics. It says it should be 12

**Comment by nikolaglumac** [comment]
@IuliaDolishniak we still don't have an endpoint for Yoroi Balance wallets so you can ignore that error message. Please test if you can restore a normal 15 word mnemonic wallet as Yoroi rewards wallet - you can generate a phrase here:  https://iancoleman.io/bip39/

File: source/renderer/app/components/wallet/WalletRestoreDialog.js [hunk 14]
```diff
@@ -551,6 +631,21 @@ export default class WalletRestoreDialog extends Component<Props, State> {
     return this.state.walletType === WALLET_RESTORE_TYPES.LEGACY;
   }
 
+  isYoroiLegacy() {
+    return this.state.walletType === WALLET_RESTORE_TYPES.YOROI_LEGACY;
+  }
+
+  isYoroi() {
```
Comment (tomothespian) [inline source/renderer/app/components/wallet/WalletRestoreDialog.js]: @DeeJayElly 
Use `isYoroiLegacy()` and `isYoroiRegular()`

Comment (DeeJayElly) [reply source/renderer/app/components/wallet/WalletRestoreDialog.js]: Fixed

**Comment by gnpf** [comment]
@DeeJayElly When selecting _Rewards Wallets_ on the tab _Yoroi_ it switches to _Daedalus_ tab and maintains the radio button selected.
It looks confusing to me, but if this is expected, the _Daedalus_ tab is not even selected after it switches.

**Comment by nikolaglumac** [comment]
@gnpf thanks for raising this up - this definitely looks like a bug with the tabs navigation - as if it renders wrong forms on click on the radio buttons. @DeeJayElly hope you can easily fix it.

**Comment by DmitriiGaico** [comment]
> @gnpf thanks for raising this up - this definitely looks like a bug with the tabs navigation - as if it renders wrong forms on click on the radio buttons. @DeeJayElly hope you can easily fix it.

We are trying to reproduce this in the latest build. It might be a false alatm. Going to report updates shortly.

**Comment by gnpf** [comment]
@nikolaglumac @DeeJayElly @DmitriiGaico I tested this again and couldn't reproduce it.

**Comment by gnpf** [comment]
@DeeJayElly In Japanese, the word _Daedalus_ is translated in the first tab, and it shouldn't be.
@nikolaglumac approved this change, can you check this, please?
 
image

**Comment by nikolaglumac** [comment]
@DeeJayElly I confirm what @gnpf just reported. I guess someone translated the word "Daedalus" by mistake.

File: source/renderer/app/components/wallet/WalletRestoreDialog.js [hunk 15]
```diff
@@ -424,15 +451,15 @@ export default class WalletRestoreDialog extends Component<Props, State> {
           skin={InputSkin}
         />
 
-        {!this.isCertificate() && (
+        {!this.isCertificate() && !this.isYoroi() && (
```
Comment (nikolaglumac) [inline source/renderer/app/components/wallet/WalletRestoreDialog.js]: @DeeJayElly please replace:
`!this.isCertificate() && !this.isYoroi()`
with:
`this.isRegular() && this.isLegacy()`

File: source/renderer/app/components/wallet/WalletRestoreDialog.js [hunk 16]
```diff
@@ -486,10 +569,7 @@ export default class WalletRestoreDialog extends Component<Props, State> {
           }
           placeholder={
             !this.isCertificate()
-              ? intl.formatMessage(messages.recoveryPhraseInputHint, {
-                  numberOfWords:
-                    walletType === WALLET_RESTORE_TYPES.LEGACY ? '12' : '15',
```
Comment (nikolaglumac) [inline source/renderer/app/components/wallet/WalletRestoreDialog.js]: @DeeJayElly we need to keep this.

File: source/renderer/app/components/wallet/WalletRestoreDialog.js [hunk 17]
```diff
@@ -486,10 +569,7 @@ export default class WalletRestoreDialog extends Component<Props, State> {
...
-                })
+              ? intl.formatMessage(messages.recoveryPhraseInputHint)
               : intl.formatMessage(messages.shieldedRecoveryPhraseInputHint, {
                   numberOfWords: 27,
```
Comment (nikolaglumac) [inline source/renderer/app/components/wallet/WalletRestoreDialog.js]: @DeeJayElly please replace `27` hardcoded value with the `PAPER_WALLET_RECOVERY_PHRASE_WORD_COUNT `.

Comment (DeeJayElly) [reply source/renderer/app/components/wallet/WalletRestoreDialog.js]: Done

File: source/renderer/app/components/wallet/WalletRestoreDialog.js (hunk 16 above)
Comment (DeeJayElly) [reply source/renderer/app/components/wallet/WalletRestoreDialog.js]: We don't need it for placeholder as the placeholder is changed now in the new design https://app.zeplin.io/project/57acfaa7436a750316161a8b/screen/5de4e5372efdcb7bb478f975

File: source/renderer/app/components/wallet/WalletRestoreDialog.js (hunk 15 above)
Comment (DeeJayElly) [reply source/renderer/app/components/wallet/WalletRestoreDialog.js]: Fixed

**Comment by DeeJayElly** [comment]
> @DeeJayElly I confirm what @gnpf just reported. I guess someone translated the word "Daedalus" by mistake.

Fixed

**Comment by daniloprates** [comment]
Do not merge this PR before `DDW-1156` is done.
